### DISCUSSION TITLE
Viewer Search - First Char

### DISCUSSION THREAD

**Discussion by MKadaner** [FarGroup/FarManager#614] [General]
https://github.com/FarGroup/FarManager/blob/8891d36fedbbbedb2ee15c82b75bec9943786d90/far/FarEng.hlf.m4#L2855-L2856

Do we want to support this feature after refactoring Viewer to use unified Search dialog? I've never used it but who am I to judge?

**Comment by yegor-mialyk** [comment]
I think it can be achieved with macro. So I'm personally for the unification.

**Comment by MKadaner** [comment]
Thanks for chiming in, @yegor-mialyk. To clarify, it's not the matter of unifying or not. I can add this functionality (I hope) to the unified dialog, The question is whether I can reduce complexity by dropping this functionality if it is not really popular.

**Comment by rohitab** [comment]
I've never used it either, but then I didn't know about it until know. I guess I need to spend more time going through the Help files.

I use incremental search a lot in other editors, like Visual Studio. This feature could be more useful if it worked like incremental search, i.e. without the dialog.

If I had a vote I would vote _No_, to reduce complexity and make things easier for you. :smile:

**Comment by MKadaner** [comment]
Thanks. Yeah, incremental search is sometimes convenient, though I do not use it often even where it works. Anyway, it's a completely different story. I think I've seen a request for incremental search in Far elsewhere. I believe it was before GitHub times. Maybe worth creating a feature request here for proper tracking.

The main advantage of removing unused features is improving maintainability. Implementing something once maybe tedious but doable. Having to keep it working while adding other, more useful features, this is where it gets nasty.

**Comment by qlost** [comment]
Ребят, сорян что по-русски, но, думаю, вы язык вспомните...
Т.е. 3 чела отменили фичу, которой +/- 10 лет, в продукте, которым пользуются тысячи. А 4й аппрувнул и смержил в мастер. И даже не предоставили workaround, выражаясь по-вашему.
Не знаю, слышали ли про Кровавый Ынтерпрайз, но в продуктах такого масштаба так не делается.

**Comment by yegor-mialyk** [comment]
I'm not an expert in LUA, but here is a sample that works for me:

```
Macro {
  description="First Char to search";
  area="Viewer";
  key="/[a-z0-9]/";
  action=function()
    Keys("F7 akey")
  end;
}
```

**Comment by MKadaner** [comment]
Воркараунд предложен выше Егором.

> Кровавый Ынтерпрайз

Не, не слышал. Чё енто?

> ... в продукте, которым пользуются тысячи ...

А сколько из них пользуются этой фичóй? Сколько вообще о ней знают?

Если серьёзно, is your grievance substantial, или просто заклеймить типичных представителей Кровавого Ынтерпрайзу?

**Reply by qlost** [reply]
> А сколько из них пользуются этой фичóй? Сколько вообще о ней знают?

Этот вопрос точно по адресу? А ответ вас действительно интересует?
А Егору спасибо. Когда обновлю версию - проверю решение.

**Comment by MKadaner** [comment]
> Do you really want the answer?

Yes. If this feature is _actually_ used by one percent of the population, then it's better to publish Egor's macro, maybe even include it into the distribution, and avoid complicating the implementation (which in this place, as Mr. Alabuzhev gently put it, is rather convoluted.) If it's half of the population, it better works out of the box.

Let me ask in my turn, did you really need this answer? Did I tell you something new?

P.S. Regarding whether I asked the right person, well, it's a public discussion and as such, the question is not to you only. I started this discussion with the express purpose of collecting opinions, to poll the audience so to say.

---
Если кому родной язык...

> А ответ вас действительно интересует?

Да. Если _реально_ пользуется этой фичей один процент населения, то можно опубликовать Егоров макрос, может быть, включить его в дистрибуцию, и не усложнять код (который в этом месте, как аккуратно высказался г. Алабужев, довольно запутанный). Если это половина населения, то лучше, чтобы это работало из коробки.

Спрошу в свою очередь, Вам действительно нужен был этот ответ? Я сообщил что-то новое?

P.S. Насчёт "по адресу", это публичная дискуссия, так что вопрос обращён не только к Вам. Я начал эту дискуссию специально для того, чтобы собрать мнения, проголосовать, так сказать.

**Comment by ditzyalbatross** [comment]
I for one have used this quite a lot since forever and I'm somewhat perplexed that it just got the unceremonious boot all of a sudden. If the macro is to be considered an adequate substitute, then it has to accept any character, just the way the original feature did.

**Reply by MKadaner** [reply]
Noted. Thanks.

**Comment by w17** [comment]
```
Macro {
  area="Viewer"; key="/\\S/";
  action=function() Keys("F7 AKey") end;
}
```
Лучше подойдёт. http://forum.ru-board.com/topic.cgi?forum=5&topic=51024&start=680#14

**Comment by w17** [comment]
Код 'запутанный', но выкидывание несколькихз строчек не сделало его проще.
Особых причин для того чтобы её выпиливать не было.
Хотя я про неё вспомнил, только когда увидел что её отменили.
Ну и макросом действительно возвращается достаточно просто.

**Comment by MKadaner** [comment]
Сейчас верну взад. Ну, не сейчас, конечно, но на выходных сделаю.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by w17** [comment]
Да не горит и я как бы даже не настаиваю.
Но лучше о таком хотя бы на форуме объявлять.
Там тоже посещаемость невелика, но больше чем здесь.

**Comment by MKadaner** [comment]
[Writing in English for the benefit of other people here.]

1. See comments in #625.
2. Re: forum: I have real difficulties monitoring more than one place for a certain topic. Some kind of mental limitation. Sorry about that. The development happens here, so I use GH for discussions and announcements. Thank you for posting the link on ru-board, by the way.
3. Re: convoluted code: Looking back I am glad I dropped this feature from the main PR. Even though it's literally less than 50 lines, it would add unproportional complexity to the PR which already was larger than it's comfortable for review.

**Comment by HamRusTal** [comment]
The ultimate macros that restore the removed functionality:
```lua
-- See https://forum.farmanager.com/viewtopic.php?p=173468#p173468
Macro {
  area="Viewer"; key="/\\S/";
  action=function() Keys("F7 AKey") end;
}
Macro {
  area="Viewer"; key="CtrlV ShiftIns";
  action=function() Keys("F7 AKey") end;
}
```
Ideally, we should have these included in the package and mentioned in the changelog, so that people like me who used the functionality aren't left guessing.

**Reply by MKadaner** [reply]
Thank you for posting it here. BTW, could they be collapsed into a single macro like this (sorry, I am not fluent with Far macros)?
```Lua
Macro {
  area="Viewer"; key="/\\S/ CtrlV ShiftIns";
  action=function() Keys("F7 AKey") end;
}
```

@alabuzhev, do you want me to make PR with it?

**Reply by HamRusTal** [reply]
AFAIK Far doesn't support mixing regexps and plain text keys. These macros can be merged this way (plus description and formatting changes I prefer personally):
```lua
-- Based on https://forum.farmanager.com/viewtopic.php?p=173468#p173468
Macro {
  description = "Bring up the Search dialog with the typed character or pasted text";
  area = 'Viewer'; key = '/\\S|[LR]CtrlV|ShiftIns/';
  action = function() Keys'F7 AKey' end
}
```
Yes, this is shorter yet more complex. Which code to choose is up to Far maintainers.

**Comment by alabuzhev** [comment]
> do you want me to make PR with it?

Would be nice, thanks.
These commits can be used as a reference: c2a2418c7ae26119d8bd63f8d88929ee4619e0c4, 3518eaf0ffee8edce3833e5ecde6cdb38ec099a0.

**Reply by MKadaner** [reply]
Will do. Thanks for the pointers.

**Comment by MKadaner** [comment]
Thank you all, See #625. Interestingly, we already had [View.ShiftInsSearch.lua](https://github.com/FarGroup/FarManager/blob/master/extra/Addons/Macros/View.ShiftInsSearch.lua) (with slightly different behavior).
